Support Create_time and Update_time in MyRocks table status

spetrunia opened the pull request:
- Create_time is read from frm file, like InnoDB. Prior to this diff, create_time was NULL.
- Update_time is in-memory only (like in InnoDB).

spetrunia commented:
Addressed the review input. Ready for another review pass.

spetrunia commented:
Take-aways from the last discussion:

- Put the timestamp into the DDL_ENTRY_INDEX_START_NUMBER.
- This will require increasing the format version for that record. Old servers will not be able to start on a datadir where a new server has run a DDL.
- In order to allow for reversible upgrades, introduce a variable similar to innodb_file_format.

spetrunia commented:
Suggestions for the variable:

The variable will be named `rocksdb_file_format`, type integer.
It will be a global read-only variable, only settable in my.cnf/server args.

* Current data format is denoted by rocksdb_file_format=1.
* The data format after this patch is rocksdb_file_format=2.
* The default setting after this patch will be `rocksdb_file_format=2`.  (Those willing to do an un-doable upgrade will have to manually set rocksdb_file_format=1 in my.cnf)

spetrunia commented:
... Implemented the variable, and switched the patch back to the "variant 2" of the on-disk format.  The PR is ready for review.

hermanlee commented on storage/rocksdb/rdb_datadic.h:
```diff
@@ -506,7 +523,10 @@ class Rdb_key_def {
   // if changing schema layout
   enum {
-    DDL_ENTRY_INDEX_VERSION = 1,
+    DDL_ENTRY_INDEX_VERSION_1 = 1,
```
These enums should probably reference ROCKSDB_DATADIC_FORMAT_INITIAL and ROCKSDB_DATADIC_FORMAT_CREATE_TIMESTAMP.

spetrunia replied:
Rephrased.

hermanlee replied:
Does rocksdb_table_dictionary_format also control the version of CF_DEFINITION, BINLOG_INFO_INDEX_NUMBER, DDL_DROP_INDEX_ONGOING version?

Or is it mainly to control the DDL_ENTRY? If it's to control DDL_ENTRY, then DDL_ENTRY_INDEX_VERSION_1 can be set directly to ROCKSDB_DATADIC_FORMAT_INITIAL.

hermanlee commented on storage/rocksdb/rdb_datadic.h:
```diff
@@ -203,6 +203,22 @@ enum {
+  rocksdb_table_dictionary_format.
+  This way, one can keep the on-disk data backward-compatible.
+*/
+const uint ROCKSDB_DATADIC_FORMAT_INITIAL = 1;
```
Maybe rename FORMAT_INDEX_ONLY?

spetrunia replied:
I don't mind renaming, but I don't understand the reasoning behind the name?

(Before this patch, all data dictionary records had version =1 so I thought _INITIAL was an obvious choice of name...)

hermanlee replied:
Initial is fine. I thought index_only would be more descriptive when compared to format_create_timestamp.

hermanlee commented on storage/rocksdb/rdb_datadic.cc:
```diff
@@ -4015,27 +4028,52 @@ bool Rdb_ddl_manager::init(Rdb_dict_manager *const dict_arg,
+      if (real_val_size < Rdb_key_def::TABLE_CREATE_TIMESTAMP_SIZE) {
+        // NO_LINT_DEBUG
+        sql_print_error( "RocksDB: DDL ENTRY V2 doesn't have timestamp");
+        delete tdef;
```
The other error return paths should probably delete tdef too for correctness, though if rocksdb doesn't load, the server terminates anyway. It might prevent asan leak issues though.

spetrunia replied:
Added `delete tdef` to other paths.

hermanlee commented on storage/rocksdb/rdb_datadic.cc:
```diff
@@ -3514,8 +3514,21 @@ bool Rdb_tbl_def::put_dict(Rdb_dict_manager *const dict,
                 m_key_count * Rdb_key_def::PACKED_SIZE * 2);
-  rdb_netstr_append_uint16(&indexes, Rdb_key_def::DDL_ENTRY_INDEX_VERSION);
+
+  if (rocksdb_table_dictionary_format <
```
It might be better if this was a switch:

```
switch (rocksdb_table_dictionary_format) {
  case ROCKSDB_DATADIC_FORMAT_INITIAL:
    rdb_netstr_append_uint16(&indexes, Rdb_key_def::DDL_ENTRY_INDEX_VERSION_1);
    set_create_time(0);
    break;
  case ROCKSDB_DATADIC_FORMAT_CREATE_TIMESTAMP:
    rdb_netstr_append_uint16(&indexes, Rdb_key_def::DDL_ENTRY_INDEX_VERSION_2);
    rdb_netstr_append_uint64(&indexes, create_time);
    break;
  default:
    DBUG_ASSERT(0);
    rdb_netstr_append_uint16(&indexes, Rdb_key_def::DDL_ENTRY_INDEX_VERSION_1);
    set_create_time(0);
    break;
}
```

This would be easier to extend. At minimum for the else {}, we should assert rocksdb_table_dictionary_format == ROCKSDB_DATADIC_FORMAT_CREATE_TIMESTAMP for this build.

spetrunia replied:
Done

hermanlee commented on storage/rocksdb/rdb_datadic.h:
```diff
@@ -203,6 +203,22 @@ enum {
+
+// Maximum possible value:
+const uint ROCKSDB_DATADIC_FORMAT_MAX = 2;
+const uint ROCKSDB_DATADIC_FORMAT_DEFAULT = ROCKSDB_DATADIC_FORMAT_MAX;
```
Can the default be ROCKSDB_DATADIC_FORMAT_INITIAL instead so that MyRocks and PerconaRocks don't automatically change behaviors? Can you commit a separate change to update it to 2 on the MariaDB branch?

hermanlee commented on mysql-test/suite/rocksdb/t/show_table_status.test:
```diff
@@ -62,3 +62,80 @@ CREATE TABLE `t1_new..............................................end`(a int) en
+--echo # MDEV-17171: Bug: RocksDB Tables do not have "Creation Date"
+--echo #
+use test;
+create table t1 (a int) engine=rocksdb;
```
Add a test for partitioned tables?

spetrunia replied:
Done

hermanlee reviewed (commented):
Looks good overall. Just a few minor comments.

spetrunia commented:
@hermanlee, I have now addressed all input except the question about rename to FORMAT_INDEX_ONLY about which I have a question posted above.

hermanlee reviewed (approved):
Code looks fine from the functional perspective. It's fine to be checked in.

lth commented:
I don't see timestamp being preserved in `::truncate` so I assume that we update create time on every truncate. Is that expected behaviour?

At least on InnoDB, it seems like truncate will not update timestamp (but InnoDB just returns filesystem ctime of the .frm file).

https://github.com/facebook/mysql-5.6/blob/9b7f226385031c3237977b9ec2971cb883b6e50a/storage/innobase/handler/ha_innodb.cc#L12535-L12537

I also wonder if just checking the .frm ctime is a simpler diff since we won't have to change the data dictionary format. I understand it's not perfect though (eg. other things could change file metadata, or maybe fs does not provide it).

george-lorch commented:
@lth checking .frm file might be OK here in 5.6, but is not portable to 8.0, so I would caution against it.

hermanlee commented:
@georgelorchpercona I think 8.0 intends to store this in the new data dictionary though. There is a FIELD_CREATED "created TIMESTAMP NOT NULL" field in the table definition for table.

george-lorch commented:
@hermanlee correct, there is a facility within 8.0 to handle this.  Create time is handled in the DD, but update an check times seem to need engine participation (still on our 8.0 MyRocks todo to figure this out).  My point, for whatever it is worth to FB MyRocks, is that basing anything on .frm is throw away for 8.0 and beyond.

lth commented:
@georgelorchpercona My point is that both solutions 1. checking .frm files, and 2. changing the engine data dictionary format to include creation time is throwaway work for 8.0+ because new data dictionary handles it for us.

Given that this is temporary, I feel like checking .frm files is a much simpler solution for 5.6

hermanlee commented:
@georgelorchpercona Yes, update time support is still needed for this patch, but it doesn't require any data dictionary changes. Since create time is supported by DD, any change we make here is considered throw away. If that is the case, it would probably be better not to modify the data dictionary and do what InnoDB does, query the .frm file. When that is ported to 8.0, the code gets removed.

spetrunia committed:
Support Create_time and Update_time in MyRocks table status

(Variant #7 of the patch) The implementation follows InnoDB:
- Create_time is taken from the .frm file creation timestamp
- Update_time is maintained in memory only and is set NULL on server
  restart

spetrunia commented:
Ok, updated the patch:
- Create_time is now taken from the .frm file's timestamp. This means that the patch doesn't make any changes to the on-disk data format anymore.
- (followup to @lth 's comment) TRUNCATE TABLE now keeps the Create_time intact.  Update_time is set to NULL (or should I set it to the current time?)  (interesting: in InnoDB `TRUNCATE TABLE` does not change the Update_time at all: https://gist.github.com/spetrunia/7b4125ea4be4e365db20565e4ce445a8)  
- Made some cleanup in the testcase
 
And the patch is now ready for another review round.

spetrunia commented:
... Wait, there will be more changes

spetrunia commented:
Now, ready for another pass.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -7549,6 +7576,7 @@ int ha_rocksdb::create_table(const std::string &table_name,
 
   /* Create table/key descriptions and put them into the data dictionary */
   m_tbl_def = new Rdb_tbl_def(table_name);
+  //m_tbl_def->set_create_time(time(nullptr));
```
Most of these changes can be removed?

spetrunia replied:
Yes, now done.  (they were actually gone in MariaDB version. I'm sorry for mixing something up and still having them there)

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -7696,8 +7726,9 @@ int ha_rocksdb::create(const char *const name, TABLE *const table_arg,
   }
-
-  DBUG_RETURN(create_table(str, table_arg, create_info->auto_increment_value));
+  int rc = create_table(str, table_arg, create_info->auto_increment_value);
```
This change can be removed?

spetrunia replied:
Yes, done.

spetrunia commented:
@hermanlee , ok, now all input is addressed.
